utility/session.go

```diff
@@ -1,157 +1,266 @@
+
+	if !slices.Contains(app.Chains, s.session.RelayChain) {
+		return fmt.Errorf("application %s does not stake for relay chain %s", app.Address, s.session.RelayChain)
+	}
+
+	if !(app.PausedHeight == -1 && app.UnstakingHeight == -1) {
+		return fmt.Errorf("application %s is either unstaked or paused", app.Address)
+	}
+
+	// TODO(#697): Filter by geo-zone
+
+	// INVESTIGATE: Consider what else we should validate for here (e.g. Application stake amount, etc.)
+
+	return nil
 }
 
-// use the seed information to determine a SHA3Hash that is used to find the closest N actors based
-// by comparing the sessionKey with the actors' public key
-func (s *session) getSessionId() ([]byte, types.Error) {
+// hydrateSessionID use both session and on-chain data to determine a unique session ID
+func (s *sessionHydrator) hydrateSessionID() error {
 	sessionHeightBz := make([]byte, 8)
-	binary.LittleEndian.PutUint64(sessionHeightBz, uint64(s.height))
+	binary.LittleEndian.PutUint64(sessionHeightBz, uint64(s.session.SessionHeight))
 
-	blockHashBz := []byte("get block hash bytes at s.sessionHeight from persistence module")
+	prevHashHeight := int64(math.Max(float64(s.session.SessionHeight)-1, 0))
+	prevHash, err := s.readCtx.GetBlockHash(prevHashHeight)
+	if err != nil {
+		return err
+	}
+	prevHashBz, err := hex.DecodeString(prevHash)
+
+	appPubKeyBz := []byte(s.session.Application.PublicKey)
+	relayChainBz := []byte(string(s.session.RelayChain))
+	geoZoneBz := []byte(s.session.GeoZone)
+
+	s.sessionIdBz = concat(sessionHeightBz, prevHashBz, geoZoneBz, relayChainBz, appPubKeyBz)
+	s.session.Id = crypto.GetHashStringFromBytes(s.sessionIdBz)
+
+	return nil
+}
 
-	appPubKey, err := crypto.NewPublicKey(s.application.GetPublicKey())
+// hydrateSessionServicers finds the servicers that are staked at the session height and populates the session with them
+func (s *sessionHydrator) hydrateSessionServicers() error {
+	// number of servicers per session at this height
+	numServicers, err := s.readCtx.GetIntParam(types.ServicersPerSessionParamName, s.session.SessionHeight)
 	if err != nil {
-		return nil, types.ErrNewPublicKeyFromBytes(err)
+		return err
 	}
 
-	return concat(sessionHeightBz, blockHashBz, []byte(s.geoZone), []byte(s.relayChain), appPubKey.Bytes()), nil
+	// returns all the staked servicers at this session height
+	servicers, err := s.readCtx.GetAllServicers(s.session.SessionHeight)
+	if err != nil {
+		return err
+	}
+
+	// OPTIMIZE: Consider updating the persistence module so a single SQL query can retrieve all of the actors at once.
+	candidateServicers := make([]*coreTypes.Actor, 0)
+	for _, servicer := range servicers {
+		// Sanity check the servicer is not paused, jailed or unstaking
+		if !(servicer.PausedHeight == -1 && servicer.UnstakingHeight == -1) {
+			return fmt.Errorf("hydrateSessionServicers should not have encountered a paused or unstaking servicer: %s", servicer.Address)
+		}
+
+		// TODO(#697): Filter by geo-zone
```

Comment:
Nice call out :sunglasses::+1: 

nit: 
```suggestion
		// TECHDEBT(#697): Filter by geo-zone
```